### PR TITLE
`lib/uksched`: New scheduling API

### PR DESCRIPTION
**Pull request by skuenzer**
This PR re-factors the `lib/uksched`-API. Main purpose is the API simplifacation and removing features that aren't needed and would be covered by thread API libraries (like `pthread-embedded`, `musl`). The updated API is strictly separating thread creation from scheduling: threads can be created in various ways in terms of allocation ownership. Afterwards, they need to be assigned to a scheduler instance for getting executed. The idea of the API is also that any thread can also be unassigned and re-assigned to a different instance.

This PR currently does not support blocking thread garbage collection callbacks and blocking thread destructor functions. `lib/ukschedcoop`'s idle thread would not be non-blocking anymore which would break the scheduler implementation.

GitHub-Depends-On: #535

### CODE REVIEW

**Review by dragosargint — APPROVED** [review]
The PR looks good! Here are some tests that I used for this PR [1]
[1] https://github.com/dragosargint/test

File: lib/posix-futex/futex.c
Comment (adinasm) [inline lib/posix-futex/futex.c]: This comment is no longer needed.

File: lib/uksched/include/uk/sched.h
Comment (adinasm) [inline lib/uksched/include/uk/sched.h]:
```suggestion
/* Exits the current thread context */
```

File: lib/uksched/include/uk/sched.h
Comment (adinasm) [inline lib/uksched/include/uk/sched.h]:
```suggestion
/* Exits the current thread and runs the given routine when releasing the
```

File: lib/uksched/include/uk/sched.h
Comment (adinasm) [inline lib/uksched/include/uk/sched.h]:
```suggestion
/* Terminates another thread */
```

File: lib/uksched/include/uk/tcb_impl.h
Comment (adinasm) [inline lib/uksched/include/uk/tcb_impl.h]:
```suggestion
 * @param tcb Reference to reserved space to custom thread control block
```

**Review by adinasm — COMMENTED** [review]
Looks good to me, I only left some minor comments.

**Comment by craciunoiuc** [comment]
Hey @skuenzer

Please check the checkpatch results + CodeFactor + Codacy and fix what's fixable

**Comment by skuenzer** [comment]
> Hey @skuenzer
> 
> Please check the checkpatch results + CodeFactor + Codacy and fix what's fixable

I am already on it 😉

**Comment by skuenzer** [comment]
Hi, thanks a lot for all the comments, I updated this PR and applied your suggestions. I saw that Codacy is reporting 4 issues but IMO these are false positives (becuase of nested macros and a macro that expands to a for-loop).

**Review by razvand — COMMENTED** [review]
This is OK with me. I'll wait for comments from @adinasm, @dragosargint and @craciunoiuc and then mark it as approved and add the reviewed and approved tags.

@skuenzer , as with PR #535 , I don't know what causes [`UK_ASSERT` Codacy issues](https://github.com/unikraft/unikraft/pull/564/checks).

File: lib/ukschedcoop/schedcoop.c
Comment (razvand) [inline lib/ukschedcoop/schedcoop.c:85]: This Codacy warning should be ignores. `min_wakeup_time` is `0` only the first time. Then it will be set to `thread->wakeup_time`.

**Review by dragosargint — APPROVED** [review]
I retested this PR and apart from the comments left by @razvand, I found no other problems.
Let it shine!
Reviewed-by: Dragos Iulian Argint <dragosargint21@gmail.com>

File: plat/common/arm/thread_start64.S
Comment (craciunoiuc) [inline plat/common/arm/thread_start64.S]: Why is this commented and not just removed?

Will it be used in the future?

File: arch/arm/arm64/ctx.S
Comment (craciunoiuc) [inline arch/arm/arm64/ctx.S]:
```suggestion
 * Copyright (c) 2022, OpenSynergy GmbH. All rights reserved.
```

File: lib/uksched/include/uk/sched.h
Comment (craciunoiuc) [inline lib/uksched/include/uk/sched.h]:
```suggestion
 * Copyright (c) 2022, NEC Laboratories GmbH, NEC Corrporation.
 *                     All rights reserved.
```

File: lib/uksched/thread.c
Comment (craciunoiuc) [inline lib/uksched/thread.c:289]: Is there a convention to how function arguments should be structured?

For example, I added a newline only when passing 80 characters.

File: lib/ukschedcoop/Config.uk
Comment (craciunoiuc) [inline lib/ukschedcoop/Config.uk]:
```suggestion
	select LIBUKSCHED

```

**Review by craciunoiuc — CHANGES_REQUESTED** [review]
Hey @skuenzer,

Some final touches from my side, I will approve after this.

`lib/ukschedcoop: Re-factor scheduler instance creation and instance object` -> This commit message is a bit too long unfortunately 😢
`lib/uksched: Call termination callbacks when thread transitions to exited` -> also this one
`lib/uksched: Provide GC hook on thread exit with uk_sched_thead_exit2()` -> also this one
`lib/ukboot, lib/uksched: Scheduler initialization inline with lib/ukalloc` -> also this one

File: lib/ukschedcoop/Config.uk
Comment (craciunoiuc) [reply lib/ukschedcoop/Config.uk]: Should there be a newline here?

File: plat/common/arm/thread_start64.S
Comment (skuenzer) [reply plat/common/arm/thread_start64.S]: I agree, this is weird. I removed it. I did this as part of PR #535. Because this PR depends on #535, you are seeing its commits also here.

File: arch/arm/arm64/ctx.S
Comment (skuenzer) [reply arch/arm/arm64/ctx.S]: Thanks for spotting this :+1:

Comment (skuenzer) [reply arch/arm/arm64/ctx.S]: I fixed it with PR #535 .

File: lib/uksched/include/uk/sched.h
Comment (skuenzer) [reply lib/uksched/include/uk/sched.h]: Thanks and fixed :+1:

File: lib/uksched/thread.c
Comment (skuenzer) [reply lib/uksched/thread.c:289]: Yes, this is true. I do this normally, too. In fact, I have a kind of exception only with functions with a lot of arguments (don't ask me for the threshold that I am using 😉 ). I noticed that pressing as much as arguments possible into a few lines does not contribute to the readability. Here, we actually have a rather unbalanced length of each argument that I find tedious to read. I am prefer keeping it this way because of this, what do you think?

File: lib/ukschedcoop/Config.uk
Comment (skuenzer) [reply lib/ukschedcoop/Config.uk]: Yes, an ending newline should be there. I updated my editor settings now. The newline was missing originally but should be there. I fixed it :+1:

**Comment by skuenzer** [comment]
> Hey @skuenzer,
> 
> Some final touches from my side, I will approve after this.
> 
> `lib/ukschedcoop: Re-factor scheduler instance creation and instance object` -> This commit message is a bit too long unfortunately 😢 `lib/uksched: Call termination callbacks when thread transitions to exited` -> also this one `lib/uksched: Provide GC hook on thread exit with uk_sched_thead_exit2()` -> also this one `lib/ukboot, lib/uksched: Scheduler initialization inline with lib/ukalloc` -> also this one

Hey @craciunoiuc , I updated the PR (including shorter commit messages, yes it is better 😉 ). Please, have a look. I prefered keeping the argument list with `uk_thread_init_bare_fn0()` and co, the rest I fixed. Thanks a lot for your review.

File: lib/uksched/thread.c
Comment (craciunoiuc) [reply lib/uksched/thread.c:289]: Yeah, I also like it like this more. Just wanted to know if there's a convention for this, for the future.

I'm marking as resolved.

**Review by craciunoiuc — APPROVED** [review]
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

**Review by adinasm — APPROVED** [review]
This is great. Thanks!
Reviewed-by: Adina Smeu <adina.smeu@gmail.com>

**Comment by razvand** [comment]
@skuenzer , the merging of #535  is now requiring a rebase of this PR. This mostly means getting rid of all commits belonging to #535 and then seeing if there are remaining conflicts.

**Comment by skuenzer** [comment]
> @skuenzer , the merging of #535 is now requiring a rebase of this PR. This mostly means getting rid of all commits belonging to #535 and then seeing if there are remaining conflicts.

Sure, I just pushed the rebased version ✅ .
It was conflict-free because this PR is a branch that was based on the branch of PR #535 (the one that just got merged)... 🤔 Unfortunately, GitHub could not detect that commits of this PR were already integrated. I am not sure if there is a different way how we could handle these "stacked PRs" with GitHub... 🤷🏻‍♂️ but it seems we rely on this manual step after a merge.

**Review by razvand — APPROVED** [review]
Thanks, @skuenzer 

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**
